Detangle Python `krb5` dependency mess

de11n opened the pull request:
## Description of changes

PyPI has a package called `krb5` ([here](https://pypi.org/project/krb5/)) that was *not* packaged in Nixpkgs. Nixpkgs also has a top-level system package called `krb5`.

`pyspnego` was adding `krb5` to its optional dependencies, but it was incorrectly referring to the top-level systems `krb5` when it *meant* the PyPI package `krb5`. Several other packages made the same mistake because they use `pyspnego[kerberos]` as optional dependencies as well.

To resolve this, I...
  * Disambiguated Python packages that *mean* the top-level systems `krb5` by having them use `pkgs.krb5` (is there a better way to do this?)
  * Refactored packages that copy/pasted `pyspnego[kerberos]`'s optional dependencies to instead use `pyspnego.optional-dependencies.kerberos` directly. This preserves the intent better and makes the fix apply to those places as well.
  * Added `krb5` the *PyPI* package to `python-modules`.
  * Fixed `requests-kerberos`'s hack to avoid this problem, now that it's solved.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
python3Packages.pykerberos: disambiguate krb5

PyPI also has a package called krb5 and this package wants the system krb5, not the PyPI one.

Commit:
python3Packages.pypsrp: propagate kerberos deps from pyspnego

This is more faithful to the project's pyproject.toml, which specifies pyspnego[kerberos] as the dependencies for kerberos.

Commit:
python3Packages.cccolutils: disambiguate krb5

PyPI also has a package called krb5 and this package wants the
system krb5, not the PyPI one.

Commit:
python3Packages.requests-credssp: propagate kerberos deps from pyspnego

This is more faithful to the project's pyproject.toml, which
specifies pyspnego[kerberos] as the dependencies for kerberos.

Commit:
python3Packages.k5test: disambiguate krb5

PyPI also has a package called krb5 and this package wants the
system krb5, not the PyPI one.

Commit:
python3Packages.krb5: init at 0.6.0

This causes pyspnego to get this krb5 package instead of the top-level
system krb5 package. That's actually intentional because pyspnego's
current dependency on the system package is incorrect.

Commit:
python3Packages.requests-kerberos: use pyspnego[kerberos]

This drops the hack to avoid these dependencies because PyPI's krb5 (not to be confused with the top-level system krb5) was not previously packaged in Nixpkgs.

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6511,9 +6511,7 @@ self: super: with self; {
   k-diffusion = callPackage ../development/python-modules/k-diffusion { };
 
-  k5test = callPackage ../development/python-modules/k5test {
-    inherit (pkgs) krb5 findutils;
```
This is exactly how it should be done.

dotlambda replied:
If you prefer we can use `krb5-c = pkgs.krb5`.

de11n replied:
`krb-c` seems like a decent compromise, so at least the build will fail if you don't override the input instead of trying to use the wrong thing.

de11n replied:
Done

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6661,6 +6659,8 @@ self: super: with self; {
 
   krakenex = callPackage ../development/python-modules/krakenex { };
 
+  krb5 = callPackage ../development/python-modules/krb5 {};
```
```suggestion
  krb5 = callPackage ../development/python-modules/krb5 { };
```

de11n replied:
Done

dotlambda commented on pkgs/development/python-modules/krb5/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+buildPythonPackage rec {
+  pname = "krb5";
+  version = "0.6.0";
```
```suggestion
  version = "0.6.0";
  pyproject = true;
```
and add `build-system`

de11n replied:
Done

de11n commented:
@dotlambda Thanks for the review.

What do you mean by...

> I think we should only do this if upstream depends on pyspnego[kerberos] rather than each individual package.

The packages that I changed *do* depend on `pyspnego[kerberos]` but, for some reason, their Nix dependencies were manually listing those specific dependencies instead of getting them from `pyspnego.optional-dependencies.kerberos`.

dotlambda commented:
> > I think we should only do this if upstream depends on pyspnego[kerberos] rather than each individual package.
> 
> The packages that I changed _do_ depend on `pyspnego[kerberos]` but, for some reason, their Nix dependencies were manually listing those specific dependencies instead of getting them from `pyspnego.optional-dependencies.kerberos`.

Great, then your change seems good. I just wanted to confirm.

Commit:
python3Packages.gssapi: disambiguate krb5

PyPI also has a package called krb5 and this package wants the
system krb5, not the PyPI one.

Commit:
python3Packages.requests-kerberos: drop unused pykerberos dependency

natsukium commented:
Result of `nixpkgs-review pr 333058` at a8db7f69 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>conkeyscan</li>
  </ul>
</details>
<details>
  <summary>29 packages built successfully:</summary>
  <ul>
    <li>aws-adfs (python312Packages.aws-adfs)</li>
    <li>barbicanclient (python311Packages.python-barbicanclient)</li>
    <li>glanceclient (python311Packages.python-glanceclient)</li>
    <li>heatclient (python311Packages.python-heatclient)</li>
    <li>ironicclient (python311Packages.python-ironicclient)</li>
    <li>manilaclient (python311Packages.python-manilaclient)</li>
    <li>nagstamon</li>
    <li>openstackclient (python311Packages.python-openstackclient)</li>
    <li>openstackclient-full</li>
    <li>python311Packages.atlassian-python-api</li>
    <li>python311Packages.aws-adfs</li>
    <li>python311Packages.exchangelib</li>
    <li>python311Packages.keystoneauth1</li>
    <li>python311Packages.krb5</li>
    <li>python311Packages.openstacksdk</li>
    <li>python311Packages.openstacksdk.man</li>
    <li>python311Packages.osc-lib</li>
    <li>python311Packages.python-cinderclient</li>
    <li>python311Packages.python-designateclient</li>
    <li>python311Packages.python-keystoneclient</li>
    <li>python311Packages.python-novaclient</li>
    <li>swiftclient (python311Packages.python-swiftclient)</li>
    <li>python311Packages.requests-kerberos</li>
    <li>python311Packages.trino-python-client</li>
    <li>python312Packages.atlassian-python-api</li>
    <li>python312Packages.exchangelib</li>
    <li>python312Packages.krb5</li>
    <li>python312Packages.requests-kerberos</li>
    <li>python312Packages.trino-python-client</li>
  </ul>
</details>

natsukium reviewed (approved):
The diff looks fine.

de11n commented:
I don't have merge bits so someone else will need to merge this.

dotlambda commented on pkgs/development/python-modules/krb5/default.nix:
```diff
@@ -0,0 +1,45 @@
+  };
+
+  nativeBuildInputs = [
+    cython
```
should be in `build-system`

de11n replied:
Done.

dotlambda commented on pkgs/development/python-modules/krb5/default.nix:
```diff
@@ -0,0 +1,45 @@
+  version = "0.6.0";
+  pyproject = true;
+
+  build-system = [ setuptools ];
```
We usually put this below `src`.

de11n replied:
Done.

dotlambda commented on pkgs/development/python-modules/krb5/default.nix:
```diff
@@ -0,0 +1,45 @@
+  pythonImportsCheck = [ "krb5" ];
+
+  meta = with lib; {
+    description = "Kerberos API bindings for Python";
```
```suggestion
    changelog = "https://github.com/jborean93/pykrb5/blob/v${version}/CHANGELOG.md";
    description = "Kerberos API bindings for Python";
```

de11n replied:
Done.

de11n commented:
I opened #337576 for this. Can you try it out?
